spec.html

```diff
@@ -42,8 +42,12 @@ <h1>String.prototype.replaceAll ( _searchValue_, _replaceValue_ )</h1>
   <emu-alg>
     1. Let _O_ be ? RequireObjectCoercible(*this* value).
     1. If _searchValue_ is neither *undefined* nor *null*, then
+      1. Let _isRegExp_ be ? IsRegExp(_searchString_).
+      1. If _isRegExp_ is true, then
+        1. Let _flags_ be ? ToString(? Get(_searchValue_, `"flags"`)).
```

Comment:
If flags is undefined, this should be the empty string, not the string undefined.

Reply:
How about this?

```
1. Let _flags_ be ? Get(_searchValue_, `"flags"`).
1. If Type(_flags_) is String, then
  1. If _flags_ does not contain `"g"`, throw a *TypeError* exception.
```

Let me know if you had something else in mind.

Reply:
That wouldn’t match the RegExp constructor, which tostrings everything but undefined.

I expect something like:
```
1. Let _flags_ be ? Get(_searchValue_, *”flags”*).
1. If Type(_flags_) is not Undefined,
  1. If ? ToString(_flags_) does not contain *”g”*, throw a *TypeError* exception.
```

Reply:
`"undefined"` doesn't contain `g`, so the end effect is the same. I think this falls under the same error as https://github.com/tc39/ecma262/pull/1517.

Reply:
Yes, that happens to be the case, but I think the explicitness is important - especially so as to not in any way repeat the mistake that https://github.com/tc39/ecma262/pull/1517 is attempting to make.

In other words, the committee/engines may not have felt it worth it to fix that particular bug, but that doesn't mean I want to see an avoidable bug occur ever again, in any proposal.

Reply:
Done in e796ca6b1b8b668e57534c2ee2bf71a986820d1f.

Reply:
What if the value is `null`? Or `3`? Why special case only `undefined` here to skip the check? This also gets us back into the same broken state if you're improperly subclassed a regex:

```js
const r = /t/;
Object.defineProperty(r, 'flags', { value: undefined });

'test'.replaceAll(r, ''); // => 'est'
```

Reply:
Because that’s how the RegExp constructor already operates.

Reply:
So what?

`exec` will happily except `null`. So will `split`, `search`, and all the others. The constructor is the least important precedent to follow here.

Reply:
I don't agree; I think that the constructor is the only precedent to follow.

 - the constructor [special-cases `undefined`](https://tc39.es/ecma262/#sec-regexpinitialize)
 - `exec` reads the original flags slot ([step 5](https://tc39.es/ecma262/#sec-regexpbuiltinexec)), and so doesn't apply here
 - `split` unconditionally [stringifies](https://tc39.es/ecma262/#sec-regexp.prototype-@@split) `.flags` in ES2015+, but prior, it read from the internal slot
 - [`match`](https://tc39.es/ecma262/#sec-regexp.prototype-@@match) and [`search`](https://tc39.es/ecma262/#sec-regexp.prototype-@@search) and [`replace`](https://tc39.es/ecma262/#sec-regexp.prototype-@@replace) all delegate to `exec`, see above
 - [`matchAll`](https://tc39.es/ecma262/#sec-regexp-prototype-matchall) matches split, but this was my mistake the committee isn't yet willing to fix - it should have matched the constructor

Thus, only `split` actually has this incorrect behavior prior to my involvement. I'm contending that it is a bug/omission/oversight - not intentional - that split doesn't match the constructor, and it's my mistake that matchAll doesn't. I'm not willing to see another split-like deviation enter the language.